src/clean.c

```diff
@@ -0,0 +1,394 @@
+
+static void print_help()
+{
+	fprintf(stderr,
+		"Usage:\n"
+		"	swupd clean [options]\n"
+		"\n"
+		"Remove cached data used for old updates from state directory.\n"
+		"\n"
+		"Options:\n"
+		"   --all                   Remove all the cached content not only old\n"
+		"   --dry-run               Just print files that would be removed\n"
+		"   -S, --statedir DIR      Specify alternate swupd state directory\n"
+		"   -h, --help              Display this help message\n"
+		"\n");
+}
+
+static struct {
+	int all;
+	int dry_run;
+} options;
+
+static struct {
+	int files_removed;
+} stats;
+
+static struct timespec now;
+
+static const struct option prog_opts[] = {
+	{ "help", no_argument, 0, 'h' },
+	{ "all", no_argument, &options.all, 1 },
+	{ "statedir", required_argument, 0, 'S' },
+	{ "dry-run", no_argument, &options.dry_run, 1 },
+	{ 0, 0, 0, 0 }
+};
+
+static bool parse_options(int argc, char **argv)
+{
+	int opt;
+	while ((opt = getopt_long(argc, argv, "hS:", prog_opts, NULL)) != -1) {
+		switch (opt) {
+		case '?':
+			return false;
+		case 'h':
+			print_help();
+			exit(0);
+		case 'S':
+			if (!optarg || !set_state_dir(optarg)) {
+				fprintf(stderr, "Invalid --statedir argument\n\n");
+				return false;
+			}
+			break;
+		case 0:
+			/* Handle options that don't have shortcut. */
+			break;
+		default:
+			fprintf(stderr, "Error: unrecognized option: -%c,\n\n", opt);
+			return false;
+		}
+	}
+	return true;
+}
+
+typedef bool(remove_predicate_func)(const char *dir, const struct dirent *entry);
+
+/* Remove files from path for which pred returns true.
+ * Currently it doesn't recursively remove directories.
+ *
+ * Unless --all is set, the files are also filtered by their age.
+ */
+static int remove_if(const char *path, remove_predicate_func pred)
+{
+	int ret = 0;
+	DIR *dir;
+	const int seconds = DAYS_TO_KEEP_FILES * 60 * 60 * 24;
+
+	dir = opendir(path);
+	if (!dir) {
+		return errno;
+	}
+
+	char *file = NULL;
+
+	while (true) {
+		free_string(&file);
+		ret = 0;
+
+		/* Reset errno to distinguish between a previous
+		 * failure and the end of stream. */
+		errno = 0;
+		struct dirent *entry;
+		entry = readdir(dir);
+		if (!entry) {
+			if (errno) {
+				ret = errno;
+			}
+			break;
+		}
+
+		if (!strcmp(entry->d_name, ".") || !strcmp(entry->d_name, "..")) {
+			continue;
+		}
+
+		string_or_die(&file, "%s/%s", path, entry->d_name);
+
+		struct stat stat;
+		ret = lstat(file, &stat);
+		if (ret != 0) {
+			fprintf(stderr, "couldn't access file %s: %s\n", file, strerror(errno));
+			continue;
+		}
+
+		if (!options.all) {
+			if (now.tv_sec - stat.st_mtime <= seconds) {
+				continue;
+			}
+		}
+
+		if (!pred(path, entry)) {
+			continue;
+		}
+
+		if (options.dry_run) {
+			printf("%s\n", file);
+		} else {
+			if (S_ISDIR(stat.st_mode)) {
+				ret = rmdir(file);
+			} else {
+				ret = unlink(file);
+			}
+			if (ret != 0) {
+				fprintf(stderr, "couldn't remove file %s: %s\n", file, strerror(errno));
```

Comment:
Possibly remove the word "file" from the error message?

Reply:
Done.